lib/ide/src/controller/module.rs

```diff
@@ -134,30 +142,44 @@ impl Handle {
         file_manager.touch(path.clone()).await?;
         let content = file_manager.read(path).await?;
         logger.info(|| "Parsing code");
-        let ast     = parser.parse(content,default())?;
+        let (ast,metadata) = parser.parse_file(content)?;
         logger.info(|| "Code parsed");
         logger.trace(|| format!("The parsed ast is {:?}", ast));
-        let id_map  = default();
-        let data    = Controller {location,ast,file_manager,parser,id_map,logger};
+        let id_map  = ast.id_map();
+        let data    = Controller {location,ast,file_manager,parser,id_map,metadata,logger};
         Ok(Handle::new_from_data(data))
     }
 
     /// Save the module to file.
-    pub fn save_file(&self) -> impl Future<Output=Result<(),RpcError>> {
-        let (path,mut fm) = self.with_borrowed(|data| {
-            (data.location.to_path(),data.file_manager.clone_ref())
+    pub async fn save_file(&self) -> impl Future<Output=Result<(),RpcError>> {
+        let (path,mut fm,metadata,ids) = self.with_borrowed(|data| {
+            let path = data.location.to_path();
+            let fm   = data.file_manager.clone_ref();
+            let ids  = serde_json::to_string(&data.id_map).expect(
+                "It should be possible to serialize idmap."
+            );
+            let meta = serde_json::to_string(&data.metadata).expect(
+                "It should be possible to serialize metadata."
+            );
+            (path,fm,meta,ids)
         });
-        fm.write(path,self.code())
-        // TODO [ao] here save also the id_map and metadata.
+        fm.write(path.clone(),self.code()).await;
```

Comment:
Will likely become irrelevant after applying previous comment — but mock File Manager truncates file on write, so we currently cannot incrementally write it.